src/check_bgp_protocols.py

```diff
@@ -0,0 +1,138 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import sys
+
+from argparse import ArgumentParser
+from subprocess import check_output, STDOUT
+
+
+def parse_args():
+    parser = ArgumentParser(
+        description='Check if all BIRD BGP protocols are up and established'
+    )
+    parser.add_argument('path', help='Path to the birdc/birdc6 binary')
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    birdc_path = args.path
+
+    code, reason = check_birdc_peers(birdc_path)
+
+    print(format_nagios_message(code, reason))
+    sys.exit(code)
+
+
+def check_birdc_peers(path):
+    try:
+        protocols = check_output([path, 'show', 'protocols'], stderr=STDOUT)
+    except (FileNotFoundError, PermissionError):
+        return (
+            ExitCodes.critical,
+            'The path to the birdc binary does not exist or is not executable'
+        )
+
+    if not protocols:
+        return (ExitCodes.critical, 'Could not retrieve BIRD\'s output')
+
+    parsed_protocols = parse_birdc_output(protocols.decode())
+
+    if not parsed_protocols:
+        return (ExitCodes.critical, 'Could not find any BGP protocol entries')
+
+    protocols_down = [
+        protocol['name'] for protocol in parsed_protocols
+        if protocol.get('info', '') != 'Established'
+    ]
+
+    if protocols_down:
+        return (ExitCodes.critical, 'Some protocols are down!')
+    else:
+        return (ExitCodes.ok, 'All protocols are up')
+
+
+def parse_birdc_output(protocols):
+    # Matching protocol lines from BIRD:
+    # BIRD 1.6.3 ready.
+    # name     proto    table    state  since       info
+    # direct1  Direct   master   up     2020-02-27
+    # b_af     BGP      master   up     2020-02-28  Established
+    # b_lb     BGP      master   start  18:43:46    Active        Socket: conn
+    # b_home   BGP      master   down   2020-02-29
+    # We are extracting the protocol name, state, date/time since last
+    # change and the additional information about the state of the session.
+    lines = protocols.splitlines()
+    lines = lines[2:]
+
+    tokenized_lines = [
+        [field.strip() for field in line.split(None, 5)]
+        for line in lines
+    ]
+    bgp_lines = [line for line in tokenized_lines if line[1] == 'BGP']
+
+    dicts = build_dicts(bgp_lines)
+
+    return dicts
```

Comment:
I would suggest to make it a bit different. Make `build_dicts` rather to `build_dict` and make it a generic one without depending on being for `BGP`. This way, you can easily reuse it in the future.
```suggestion
    bgp_dicts = [
        build_dict(line) for line in tokenized_lines if line[1] == 'BGP'
    ]
    
    return bgp_dicts
```

Reply:
This was simplified in a1a6d61